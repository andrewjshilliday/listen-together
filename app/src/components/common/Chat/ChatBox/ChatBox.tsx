import React, { useRef, useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { Scrollbars } from 'react-custom-scrollbars';
import { Message, ChatInput, OnlineUsers } from '../../../common';
import { ApplicationState } from '../../../../store';
import { setPopoverBoxVisibility } from '../../../../store/chat';
import styled from 'styled-components';

const useCloseChatBox = (ref: React.RefObject<HTMLDivElement>) => {
  const popoverBoxVisible = useSelector((state: ApplicationState) => state.chat.popoverChatBoxVisible);
  const dispatch = useDispatch();

  const handleClickOutside = (event: any) => {
    if (ref.current && !ref!.current.contains(event.target) && popoverBoxVisible) {
      dispatch(setPopoverBoxVisibility(false));
    }
  }

  useEffect(() => {
    document.addEventListener("mouseup", handleClickOutside);
    return () => {
      document.removeEventListener("mouseup", handleClickOutside);
    };
  });
}

const ChatBox: React.FC = (props: any) => {
  const messages = useSelector((state: ApplicationState) => state.chat.messages);
  const username = useSelector((state: ApplicationState) => state.room.username);
  const wrapperRef = useRef<HTMLDivElement>(null);
  const messagesContainer = useRef<HTMLDivElement>(null);
  useCloseChatBox(wrapperRef);

  useEffect(() => {
    const element = messagesContainer.current?.firstElementChild?.firstElementChild;
    if (element) {
      element.scrollTop = element.scrollHeight;
    }
  }, [messages]);

  return (
    <ChatBoxContainer>
      <OnlineUsers />
      <MessagesContainer ref={messagesContainer}>
        <StyledScrollbar>
          {messages.map((message, i) => <Message key={i} message={message} name={username ?? ''}/>)}
        </StyledScrollbar>
      </MessagesContainer>
      <ChatInput />
    </ChatBoxContainer>
  );
}

export default ChatBox;


const ChatBoxContainer = styled.div`
  display: flex;
  flex-direction: column;
  /* flex-grow: 1; */
  height: 100%;
  width: 100%;
`;
const MessagesContainer = styled.div`
  display: flex;
  flex-direction: column;
  flex-grow: 1;
  width: 100%;
  overflow: auto;
`;
const StyledScrollbar = styled(Scrollbars)`
  >div {
    display: flex;
    flex-direction: column;
    scroll-behavior: smooth;
  }
`;
