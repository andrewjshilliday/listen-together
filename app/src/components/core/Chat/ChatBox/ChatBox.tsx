import React, { useState, useRef, useEffect } from 'react';
import { TextField, InputAdornment, IconButton } from '@material-ui/core';
import SendIcon from '@material-ui/icons/Send';
import { Scrollbars } from 'react-custom-scrollbars';
import { useChat, useRoom } from '../../../providers';
import { Message } from '../../../core';
import styled from 'styled-components';

const useOutsideAlerter = (ref: React.RefObject<HTMLDivElement>) => {
  const chatProvider = useChat();

  const handleClickOutside = (event: any) => {
    if (ref.current && !ref!.current.contains(event.target) && chatProvider.chatBoxVisible) {
      chatProvider.actions.setVisibility(false);
    }
  }

  useEffect(() => {
    document.addEventListener("mousedown", handleClickOutside);
    return () => {
      document.removeEventListener("mousedown", handleClickOutside);
    };
  });
}

const ChatBox: React.FC = (props: any) => {
  const chatProvider = useChat();
  const roomProvider = useRoom();
  const [message, setMessage] = useState('');
  const wrapperRef = useRef<HTMLDivElement>(null);
  useOutsideAlerter(wrapperRef);

  const sendMessage = () => {
    chatProvider.actions.sendMessage(message);
    setMessage('');
  }

  return (
    <ChatBoxContainer ref={wrapperRef} visible={chatProvider.chatBoxVisible}>
      <MessagesContainer>
        <StyledScrollbar>
          {chatProvider.messages.map((message, i) => <Message key={i} message={message} name={roomProvider.username ?? ''}/>)}
        </StyledScrollbar>
      </MessagesContainer>
      <InputContainer>
        <TextField label="Type a message" value={message} onChange={(e) => setMessage(e.target.value)}
          onKeyPress={(e) => {
            if (e.key === 'Enter') {
              e.preventDefault();
              sendMessage();
            }
          }}
          InputProps={{
            endAdornment: (
              <InputAdornment position="end">
                <IconButton onClick={() => sendMessage()}>
                  <SendIcon />
                </IconButton>
              </InputAdornment>
            )
          }} />
        </InputContainer>
    </ChatBoxContainer>
  );
}

export default ChatBox;


const ChatBoxContainer = styled.div<{visible: boolean}>`
  ${props => !props.visible && `
    display: none;
  `}
  background: white;
  position: absolute;
  bottom: 40px;
  right: 25px;
  height: 500px;
  width: 500px;
  border: 1px solid black;
  border-radius: 2em;
`;
const StyledScrollbar = styled(Scrollbars)`
  >div {
    display: flex;
    flex-direction: column;
  }
`;
const MessagesContainer = styled.div`
  display: flex;
  flex-direction: column;
  width: 100%;
  height: 430px;
  overflow: auto;
`;
const InputContainer = styled.div`
  margin: 0 25px;
  border-top: 1px solid #ccc;
  >div {
    width: 100%;
    >div button {
      :hover { color: var(--primary); }
    }
  }
`;
