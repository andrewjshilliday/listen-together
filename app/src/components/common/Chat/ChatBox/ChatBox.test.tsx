import React from 'react';
import renderer from 'react-test-renderer';
import ChatBox from './ChatBox';

jest.mock('../../../providers/ChatProvider', () => ({
  useChat: () => ({
    messages: []
  })
}));

jest.mock('../../../providers/RoomProvider', () => ({
  useRoom: () => ({
    users: []
  })
}));

it('renders correctly', () => {
  const tree = renderer.create(<ChatBox />).toJSON();
  expect(tree).toMatchSnapshot();
});
