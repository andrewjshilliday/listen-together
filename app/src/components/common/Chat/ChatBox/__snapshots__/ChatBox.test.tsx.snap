// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="sc-fznxsB ebTagX"
>
  <div
    className="sc-fzoLag dEeiVn"
  >
    <span>
      Looks like you're on your own here!
    </span>
  </div>
  <div
    className="sc-fznJRM cCzPPj"
  >
    <div
      className="sc-fznWqX jqGeqS"
      style={
        Object {
          "height": "100%",
          "overflow": "hidden",
          "position": "relative",
          "width": "100%",
        }
      }
    >
      <div
        style={
          Object {
            "WebkitOverflowScrolling": "touch",
            "bottom": 0,
            "left": 0,
            "marginBottom": 0,
            "marginRight": 0,
            "overflow": "scroll",
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      />
      <div
        style={
          Object {
            "borderRadius": 3,
            "bottom": 2,
            "display": "none",
            "height": 6,
            "left": 2,
            "position": "absolute",
            "right": 2,
          }
        }
      >
        <div
          style={
            Object {
              "backgroundColor": "rgba(0,0,0,.2)",
              "borderRadius": "inherit",
              "cursor": "pointer",
              "display": "block",
              "height": "100%",
              "position": "relative",
            }
          }
        />
      </div>
      <div
        style={
          Object {
            "borderRadius": 3,
            "bottom": 2,
            "display": "none",
            "position": "absolute",
            "right": 2,
            "top": 2,
            "width": 6,
          }
        }
      >
        <div
          style={
            Object {
              "backgroundColor": "rgba(0,0,0,.2)",
              "borderRadius": "inherit",
              "cursor": "pointer",
              "display": "block",
              "position": "relative",
              "width": "100%",
            }
          }
        />
      </div>
    </div>
  </div>
  <div
    className="sc-fzoyAV kVBBCN"
  >
    <div
      className="MuiFormControl-root MuiTextField-root"
      onKeyPress={[Function]}
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
        data-shrink={false}
      >
        Type a message
      </label>
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoFocus={false}
          className="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd"
          disabled={false}
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          type="text"
          value=""
        />
        <div
          className="MuiInputAdornment-root MuiInputAdornment-positionEnd"
        >
          <button
            className="MuiButtonBase-root MuiIconButton-root"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiIconButton-label"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M2.01 21L23 12 2.01 3 2 10l15 2-15 2z"
                />
              </svg>
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
